### PR TITLE
[flink] Cdc ingestion support warkmark strategy

### PR DESCRIPTION
**Pull request by MonsterChenzhuo**
### Purpose

Cdc ingestion support warkmark strategy.

### CODE REVIEW

**Comment by MonsterChenzhuo** [comment]
@yuzelin PTAL,thanks.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        } else if (source instanceof PulsarSource) {
+            return new PulsarCdcTimestampExtractor();
+        }
+        throw new IllegalArgumentException(
+                "Unsupported source type: " + source.getClass().getName());
+    }
+
+    /** Timestamp extractor for MongoDB sources in CDC applications. */
+    public static class MongoDBCdcTimestampExtractor
+            implements CdcTimestampExtractor, Serializable {
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java:50]: done.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SynchronizationActionBase.java [hunk 2]
```diff
@@ -123,9 +130,20 @@ protected Object buildSource() {
 
     private DataStreamSource<String> buildDataStreamSource(Object source) {
         if (source instanceof Source) {
+            WatermarkStrategy<String> watermarkStrategy;
+            if (!(source instanceof KafkaSource)
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/SynchronizationActionBase.java:134]: done.

File: paimon-core/src/main/java/org/apache/paimon/utils/JsonSerdeUtil.java [hunk 3]
```diff
@@ -208,6 +208,26 @@ public static <T extends JsonNode> T asSpecificNodeType(String json, Class<T> cl
         return clazz.cast(resultNode);
     }
 
+    /**
+     * JsonExtractor is a utility class providing methods to extract data from a JSON string. It
+     * uses Jackson's ObjectMapper to parse JSON and allows for flexible data retrieval based on a
+     * specified path within the JSON structure.
+     *
+     * <p>Example: String json = "{\"data\": {\"id\": 123, \"name\": \"Test\"}}"; int id =
+     * JsonExtractor.extractValue(json, Integer.class, "data", "id");
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-core/src/main/java/org/apache/paimon/utils/JsonSerdeUtil.java:217]: done.

**Comment by MonsterChenzhuo** [comment]
@yuzelin PTAL,thanks.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /** Timestamp extractor for Pulsar sources in CDC applications. */
+    public static class PulsarCdcTimestampExtractor implements CdcTimestampExtractor {
+
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public long extractTimestamp(String record) throws JsonProcessingException {
+            return JsonSerdeUtil.extractValue(record, Long.class, "ts");
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java:111]: I think Pulsar should be the same to Kafka?

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcWatermarkStrategy.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            @Override
+            public void onEvent(String record, long timestamp, WatermarkOutput output) {
+                long tMs;
+                try {
+                    tMs = timestampExtractor.extractTimestamp(record);
+                } catch (JsonProcessingException e) {
+                    throw new RuntimeException(e);
+                }
+                currentMaxTimestamp = Math.max(currentMaxTimestamp, tMs);
+                output.emitWatermark(new Watermark(currentMaxTimestamp - 1));
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcWatermarkStrategy.java:60]: Should output `tMs`?

**Review by yuzelin — COMMENTED** [review]
Left some comments. Could you add some tests ?

File: paimon-core/src/main/java/org/apache/paimon/utils/JsonSerdeUtil.java [hunk 6]
```diff
@@ -208,6 +208,35 @@ public static <T extends JsonNode> T asSpecificNodeType(String json, Class<T> cl
         return clazz.cast(resultNode);
     }
 
+    /**
+     * JsonExtractor is a utility class providing methods to extract data from a JSON string. It
+     * uses Jackson's ObjectMapper to parse JSON and allows for flexible data retrieval based on a
+     * specified path within the JSON structure.
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-core/src/main/java/org/apache/paimon/utils/JsonSerdeUtil.java]: done.

**Comment by MonsterChenzhuo** [comment]
@yuzelin PTAL,thanks.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                return JsonSerdeUtil.extractValue(record, Long.class, "ts") * 1000;
+            } else if (JsonSerdeUtil.isNodeExists(record, "payload", "source", "connector")) {
+                // Dbz json
+                return JsonSerdeUtil.extractValue(record, Long.class, "payload", "ts_ms");
+            } else if (JsonSerdeUtil.isNodeExists(record, "source", "connector")) {
+                // Dbz json
+                return JsonSerdeUtil.extractValue(record, Long.class, "ts_ms");
+            }
+            throw new RuntimeException(
+                    "Unsupported messageQueue json format for timestamp extraction");
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java]: done.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+public class CdcTimestampExtractorFactory implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Map<Class<?>, Supplier<CdcTimestampExtractor>> extractorMap =
+            new HashMap<>();
+
+    static {
+        extractorMap.put(MongoDBSource.class, MysqlCdcTimestampExtractor::new);
+        extractorMap.put(MySqlSource.class, MongoDBCdcTimestampExtractor::new);
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/watermark/CdcTimestampExtractorFactory.java]: done.

**Review by yuzelin — APPROVED** [review]
+1

**Pull request merged**